### DISCUSSION TITLE
IServiceProvider

### DISCUSSION THREAD

**Discussion by workgroupengineering** [AvaloniaUI/Avalonia#15972] [Q&A]
The XamlX compiler allows the AvaloniaObject instance with ctor that takes a single IServiceProvider parameter.
Can I add a custom service that can be recovered from this ctor?

**Comment by thevortexcloud** [comment]
I am pretty sure this is just because WPF does the same thing. It mainly just contains internal services used by Avalonia that might be useful in custom code. Here is the WPF docs on it:

https://learn.microsoft.com/en-us/dotnet/desktop/xaml-services/service-contexts-with-type-converters-and-markup-extensions

It's not really meant for generic use. You should probably just set up DI with a static locator if you need to add custom services.

**Reply by maxkatz6** [reply]
Yes, this service provider **is not** intended for business logic.

**Reply by workgroupengineering** [reply]
> Yes, this service provider **is not** intended for business logic.

my intent is not for business logic. It's finding a common way to abstract platforms when creating a custom control while avoiding `#if #else #endif` and private APIs. [DependencyService](https://learn.microsoft.com/en-us/previous-versions/dotnet/api/xamarin.forms.dependencyservice?view=xamarin-forms) was used in Xamarin.Form. At the moment, since there is no common way to solve this problem in Avalonia, each developer uses the DI that is familiar to him and there is the great possibility of having multiple different DIs in the same app, with the obvious consequences of the case.

**Reply by thevortexcloud** [reply]
Most DIs implement the `IServiceProvider` interface, which is itself located in the System namespace. 

https://learn.microsoft.com/en-us/dotnet/api/system.iserviceprovider?view=net-8.0

I don't really see an issue with having some sort of dependency on that. That said forcing people to use DI is also not a great pattern in most cases as far as I am concerned (although if it's internal to your library it's probably not a big deal as long as the consumer does not need to deal with it). There is also nothing stopping you from writing an additional abstraction on top of `IServiceProvider` that lets you combine multiple DI containers. Splat for example already does this and has multiple adapters for different DIs despite it is itself a DI container. The idea being you can continue to use anything that depends on Splat while also customising it with what ever DI you prefer.

https://github.com/reactiveui/splat

**Reply by workgroupengineering** [reply]
Regarding the business logic I agree 1000%, but not at the level of the framework components like CameraPreview or WebBrowser can be.

Suppose CameraPreview depends on Autofact and WebBrowser depends on Splat. If due to a bug I am forced to update CameraPreview, it may automatically update Autofact as well. If Autofact has a shared dependency with Splat and this dependency is updated with Autofact, Spalt may not work correctly.
